// @hashgraph/sdk/lib/token/TokenNftInfoQuery.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Query = _interopRequireWildcard(require("../query/Query.cjs"));

var _NftId = _interopRequireDefault(require("./NftId.cjs"));

var _AccountId = _interopRequireDefault(require("../account/AccountId.cjs"));

var _TokenId = _interopRequireDefault(require("../token/TokenId.cjs"));

var _TokenNftInfo = _interopRequireDefault(require("./TokenNftInfo.cjs"));

var _Hbar = _interopRequireDefault(require("../Hbar.cjs"));

var _long = _interopRequireDefault(require("long"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * @namespace proto
 * @typedef {import("@hashgraph/proto").IQuery} proto.IQuery
 * @typedef {import("@hashgraph/proto").IQueryHeader} proto.IQueryHeader
 * @typedef {import("@hashgraph/proto").IResponse} proto.IResponse
 * @typedef {import("@hashgraph/proto").ITokenNftInfo} proto.ITokenNftInfo
 * @typedef {import("@hashgraph/proto").IResponseHeader} proto.IResponseHeader
 * @typedef {import("@hashgraph/proto").ITokenGetNftInfoQuery} proto.ITokenGetNftInfoQuery
 * @typedef {import("@hashgraph/proto").ITokenGetNftInfosQuery} proto.ITokenGetNftInfosQuery
 * @typedef {import("@hashgraph/proto").ITokenGetAccountNftInfosQuery} proto.ITokenGetAccountNftInfosQuery
 * @typedef {import("@hashgraph/proto").ITokenGetNftInfoResponse} proto.ITokenGetNftInfoResponse
 * @typedef {import("@hashgraph/proto").ITokenGetNftInfosResponse} proto.ITokenGetNftInfosResponse
 * @typedef {import("@hashgraph/proto").ITokenGetAccountNftInfosResponse} proto.ITokenGetAccountNftInfosResponse
 */

/**
 * @typedef {import("../channel/Channel.js").default} Channel
 */

/**
 * @augments {Query<TokenNftInfo[]>}
 */
class TokenNftInfoQuery extends _Query.default {
  /**
   * @param {object} properties
   * @param {NftId | string} [properties.nftId]
   * @param {AccountId | string} [properties.accountId]
   * @param {TokenId | string} [properties.tokenId]
   * @param {Long | number} [properties.start]
   * @param {Long | number} [properties.end]
   */
  constructor(properties = {}) {
    super();
    /**
     * @private
     * @type {?NftId}
     */

    this._nftId = null;

    if (properties.nftId != null) {
      this.setNftId(properties.nftId);
    }
    /**
     * @private
     * @type {?AccountId}
     */


    this._accountId = null;

    if (properties.accountId != null) {
      this.setAccountId(properties.accountId);
    }
    /**
     * @private
     * @type {?TokenId}
     */


    this._tokenId = null;

    if (properties.tokenId != null) {
      this.setTokenId(properties.tokenId);
    }
    /**
     * @private
     * @type {?Long}
     */


    this._start = null;

    if (properties.start != null) {
      this.setStart(properties.start);
    }
    /**
     * @private
     * @type {?Long}
     */


    this._end = null;

    if (properties.end != null) {
      this.setEnd(properties.end);
    }
  }
  /**
   * @internal
   * @param {proto.IQuery} query
   * @returns {TokenNftInfoQuery}
   */


  static _fromProtobuf(query) {
    if (query.tokenGetNftInfo != null) {
      const info =
      /** @type {proto.ITokenGetNftInfoQuery} */
      query.tokenGetNftInfo;
      return new TokenNftInfoQuery({
        nftId: info.nftID != null ? _NftId.default._fromProtobuf(info.nftID) : undefined
      });
    } else if (query.tokenGetAccountNftInfos != null) {
      const info =
      /** @type {proto.ITokenGetAccountNftInfosQuery} */
      query.tokenGetAccountNftInfos;
      return new TokenNftInfoQuery({
        accountId: info.accountID != null ? _AccountId.default._fromProtobuf(info.accountID) : undefined,
        start: info.start != null ? info.start : undefined,
        end: info.end != null ? info.end : undefined
      });
    } else {
      const info =
      /** @type {proto.ITokenGetNftInfosQuery} */
      query.tokenGetNftInfos;
      return new TokenNftInfoQuery({
        tokenId: info.tokenID != null ? _TokenId.default._fromProtobuf(info.tokenID) : undefined,
        start: info.start != null ? info.start : undefined,
        end: info.end != null ? info.end : undefined
      });
    }
  }
  /**
   * @returns {?NftId}
   */


  get nftId() {
    return this._nftId;
  }
  /**
   * Set the token ID for which the info is being requested.
   *
   * @param {NftId | string} nftId
   * @returns {TokenNftInfoQuery}
   */


  setNftId(nftId) {
    this._nftId = typeof nftId === "string" ? _NftId.default.fromString(nftId) : _NftId.default._fromProtobuf(nftId._toProtobuf());
    return this;
  }
  /**
   * @returns {?AccountId}
   */


  get accountId() {
    return this._accountId;
  }
  /**
   * Set the token ID for which the info is being requested.
   *
   * @param {AccountId | string} accountId
   * @returns {TokenNftInfoQuery}
   */


  setAccountId(accountId) {
    this._accountId = typeof accountId === "string" ? _AccountId.default.fromString(accountId) : _AccountId.default._fromProtobuf(accountId._toProtobuf());
    return this;
  }
  /**
   * @returns {?TokenId}
   */


  get tokenId() {
    return this._tokenId;
  }
  /**
   * Set the token ID for which the info is being requested.
   *
   * @param {TokenId | string} tokenId
   * @returns {TokenNftInfoQuery}
   */


  setTokenId(tokenId) {
    this._tokenId = typeof tokenId === "string" ? _TokenId.default.fromString(tokenId) : _TokenId.default._fromProtobuf(tokenId._toProtobuf());
    return this;
  }
  /**
   * @returns {?Long}
   */


  get start() {
    return this._start;
  }
  /**
   * Set the token ID for which the info is being requested.
   *
   * @param {Long | number} start
   * @returns {TokenNftInfoQuery}
   */


  setStart(start) {
    this._start = typeof start === "number" ? _long.default.fromNumber(start) : start;
    return this;
  }
  /**
   * @returns {?Long}
   */


  get end() {
    return this._end;
  }
  /**
   * Set the token ID for which the info is being requested.
   *
   * @param {Long | number} end
   * @returns {TokenNftInfoQuery}
   */


  setEnd(end) {
    this._end = typeof end === "number" ? _long.default.fromNumber(end) : end;
    return this;
  }
  /**
   * @override
   * @param {import("../client/Client.js").default<Channel, *>} client
   * @returns {Promise<Hbar>}
   */


  async getCost(client) {
    let cost = await super.getCost(client);

    if (cost.toTinybars().greaterThan(25)) {
      return cost;
    } else {
      return _Hbar.default.fromTinybars(25);
    }
  }
  /**
   * @override
   * @internal
   * @param {Channel} channel
   * @param {proto.IQuery} request
   * @returns {Promise<proto.IResponse>}
   */


  _execute(channel, request) {
    if (this._nftId != null) {
      return channel.token.getTokenNftInfo(request);
    } else if (this._accountId != null) {
      return channel.token.getAccountNftInfos(request);
    } else {
      return channel.token.getTokenNftInfos(request);
    }
  }
  /**
   * @override
   * @internal
   * @param {proto.IResponse} response
   * @returns {proto.IResponseHeader}
   */


  _mapResponseHeader(response) {
    let infos;

    if (this._nftId != null) {
      infos =
      /** @type {proto.ITokenGetNftInfoResponse} */
      response.tokenGetNftInfo;
    } else if (this._accountId != null) {
      infos =
      /** @type {proto.ITokenGetAccountNftInfosResponse} */
      response.tokenGetAccountNftInfos;
    } else {
      infos =
      /** @type {proto.ITokenGetNftInfosResponse} */
      response.tokenGetNftInfos;
    }

    return (
      /** @type {proto.IResponseHeader} */
      infos.header
    );
  }
  /**
   * @override
   * @internal
   * @param {proto.IResponse} response
   * @param {AccountId} nodeAccountId
   * @param {proto.IQuery} request
   * @returns {Promise<TokenNftInfo[]>}
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars


  _mapResponse(response, nodeAccountId, request) {
    let nfts = [];
    /** @type {proto.ITokenGetNftInfoResponse} */

    response.tokenGetNftInfo;

    if (this._nftId != null) {
      nfts = [
      /** @type {proto.ITokenNftInfo} */

      /** @type {proto.ITokenGetNftInfoResponse} */
      response.tokenGetNftInfo.nft];
    } else if (this._accountId != null) {
      nfts =
      /** @type {proto.ITokenNftInfo[]} */

      /** @type {proto.ITokenGetAccountNftInfosResponse} */
      response.tokenGetAccountNftInfos.nfts;
    } else {
      nfts =
      /** @type {proto.ITokenNftInfo[]} */

      /** @type {proto.ITokenGetNftInfosResponse} */
      response.tokenGetNftInfos.nfts;
    }

    return Promise.resolve(nfts.map(nft => _TokenNftInfo.default._fromProtobuf(
    /** @type {proto.ITokenNftInfo} */
    nft)));
  }
  /**
   * @override
   * @internal
   * @param {proto.IQueryHeader} header
   * @returns {proto.IQuery}
   */


  _onMakeRequest(header) {
    if (this._nftId != null) {
      return {
        tokenGetNftInfo: {
          header,
          nftID: this._nftId != null ? this._nftId._toProtobuf() : null
        }
      };
    } else if (this._accountId != null) {
      return {
        tokenGetAccountNftInfos: {
          header,
          accountID: this._accountId != null ? this._accountId._toProtobuf() : null,
          start: this._start,
          end: this._end
        }
      };
    } else {
      return {
        tokenGetNftInfos: {
          header,
          tokenID: this._tokenId != null ? this._tokenId._toProtobuf() : null,
          start: this._start,
          end: this._end
        }
      };
    }
  }

} // eslint-disable-next-line @typescript-eslint/unbound-method


exports.default = TokenNftInfoQuery;

_Query.QUERY_REGISTRY.set("tokenGetNftInfo", TokenNftInfoQuery._fromProtobuf);
